plugins/obs-filters,libobs: Add drop-shadow filter

tjienta opened the pull request:
![drop-shadow](https://user-images.githubusercontent.com/29837949/41427453-4d1cbab6-7007-11e8-9dc6-f56cc2347af9.png)
Add a drop-shadow video filter. This video filter uses the alpha
of the source to create a drop-shadow below it.
It has the following settings, all sliders:
 - Shadow color
 - Shadow opacity
 - Shadow offset X/Y
 - Shadow blur radius

This is useful for adding shadow under text that changes, or to
seperate a green-screened talking-head from the background.

For performance reasons the blur of the shadow is be done in several
passes using the Kawase Blur algorithm. These passes ping-pong two
shadow textures between each other. Up to 7 passes are needed for
a blur radius of 25 pixels, each pass only takes 4 samples (16 pixels).

The only thing required from libobs to handle multi-pass rendering
is access to the source->filter_texrender texture. Which contains
the image from a previous filter in the chain.

tjienta commented:
I am going to reverse the offset sliders, so when you move the X-slider to the right, the shadow goes to the right as well. If you move the Y-slider to the right, then shadow goes down.

RytoEX commented on libobs/obs-source.c:
```diff
@@ -1937,6 +1937,20 @@ obs_source_t *obs_filter_get_target(const obs_source_t *filter)
 
+gs_texture_t *obs_filter_get_texture(const obs_source_t *filter)
+{
+	if (!obs_ptr_valid(filter, "obs_filter_get_texrender")) {
```
Style:  Remove the curly braces for this `if` statement.

RytoEX commented on libobs/obs-source.c:
```diff
@@ -1937,6 +1937,20 @@ obs_source_t *obs_filter_get_target(const obs_source_t *filter)
+	}
+
+	if (!obs_ptr_valid(filter->filter_texrender,
+			"obs_filter_get_texrender")) {
```
Style:  Remove the curly braces for this `if` statement.

RytoEX commented on plugins/obs-filters/data/drop_shadow_filter.effect:
```diff
@@ -0,0 +1,190 @@
+		x.b < 0.081 ? low.b : high.b
+	);
+
+	// Pre multiply alpha.
```
Spelling:  Use either "premultiply" or "pre-multiply".  Other variants of "premultiply" occur throughout here, so you should probably keep them all consistent.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+	uint32_t shadow_color = (uint32_t)obs_data_get_int(settings,
+			SETTING_SHADOW_COLOR);
+
+	shadow_color&= 0x00ffffff;
```
Style:  Separate the variable name and the operator with a single space.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+			SETTING_SHADOW_COLOR);
+
+	shadow_color&= 0x00ffffff;
+	shadow_color|= ((opacity * 255) / 100) << 24;
```
Style:  Separate the variable name and the operator with a single space.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+static void *drop_shadow_create(obs_data_t *settings, obs_source_t *context)
+{
+	struct drop_shadow_filter_data *filter =
+		bzalloc(sizeof(struct drop_shadow_filter_data));
```
Style:  Use two tabs to indent this line continuation.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+		float kawase_distance)
+{
+	gs_texture_t *shadow = gs_texrender_get_texture(
+		shadow_source);
```
Style:  Use two tabs to indent this line continuation.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+ * The first pass does the following operations:
+ *  * Sample the alpha channel from the source image.
+ *  * Shift the image by an offset.
+ *  * Blur using Kawas blur.
```
Spelling:  "Kawas" should be "Kawase".

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+
+static kawase_entry_t get_kawase_entry(int blur_radius)
+{
+	if (blur_radius > 25) {
```
Style:  Remove the curly braces for this `if` statement.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+
+	kawase_entry_t kawase_entry = get_kawase_entry(filter->blur_radius);
+
+	// Start the rendering of the filter, first by processing
```
Style:  Multi-line comments should use `/* */`.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+			SWAP(gs_texrender_t *, front, back);
+		}
+
+		// Last pass will blur one more time and composit the original
```
This line and the next seem to be duplicates?

Style:  Multi-line comments should use `/* */`.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+
+		int pass = 0;
+		drop_shadow_render_first_pass(filter, back, front,
+			kawase_entry.distances[pass]);
```
Style: Use two tabs to indent this line continuation.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+		// Last pass will blur one more time and composit the original
+		// image on top of the blurred-shadow.
+		set_parameters(filter, front, 
+			kawase_entry.distances[pass]);
```
Style: Use two tabs to indent this line continuation.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,418 @@
+		set_parameters(filter, front, 
+			kawase_entry.distances[pass]);
+		obs_source_process_filter_end(filter->context, filter->effect,
+			filter->total_width, filter->total_height);
```
Style: Use two tabs to indent this line continuation.

RytoEX commented on plugins/obs-filters/data/locale/en-US.ini:
```diff
@@ -74,3 +75,9 @@ Compressor.AttackTime="Attack (ms)"
+DropShadowFilter.BlurRadius="Blur Radius"
+DropShadowFilter.Opacity="Opacity"
+DropShadowFilter.ShadowColor="Shadow Color"
+
```
There seems to be an extra newline here?

tjienta commented:
@RytoEX Thank you for the review, I have made the changes locally, I will push once I made a few more changes.

tjienta commented:
Okay I updated the pull request, removed colour conversion that should not have been done. Also reduced code duplication. And added a few more comments. Also fixed the direction of the offset sliders so that it feels more natural. I am pretty happy with it now.

RytoEX commented:
There are still three unresolved inline comments from my previous review.  Please address those.

tjienta commented:
I am not sure how I missed those, but I addressed those now.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,400 @@
+
+	kawase_entry_t kawase_entry = get_kawase_entry(filter->blur_radius);
+
+	/* Start the rendering of the filter, first by processing
```
Style:  Use this format for multi-line comments:
```
/*
 * text
 * goes
 * here
 */

Sorry, I should have been more specific in my initial request.

tjienta replied:
Done.

RytoEX commented on plugins/obs-filters/drop-shadow-filter.c:
```diff
@@ -0,0 +1,400 @@
+	float distances[7];
+} kawase_entry_t;
+
+/* In the kawase blur algorithm four texture samples are taken in
```
In comments, please capitalize "Kawase" as it is a proper name.  Variable names can follow whatever naming/capitalization convention being used (snake_case, camelCase, etc.).  Be sure to check this throughout.

tjienta replied:
Done.

RytoEX commented:
As an aside, this would address [Mantis Issue 556](https://obsproject.com/mantis/view.php?id=556).

tjienta commented:
I made the modification to use seperate texture targets for each pass to reduce graphic pipeline stalls.

tjienta commented:
Something is wrong, it looks like I no longer am able to retrieve the source image.
Did something change in how the stack of filters need to be processed?

tjienta commented:
Okay, I found some time and fixed it. It now also allocated and deallocated the texture targets based on how large the current blur setting is.

tjienta commented:
I reverted the GS_R8 texture target to GS_RGB. The plugin seems stable again, I have been running it for about  week again.

tjienta commented:
I've rebased from master since OBS 23.
I am not sure why the Mac build doesn't work, seems to be something about installing prerequisites, I don't think this is about my code.

Gol-D-Ace commented:
Just gave Travis a little poke ;)

tjienta commented:
Thanks :-)

emdemir commented:
Hi, I built master with your patch just now and everything seems to be working fine, aside from one small nitpick: When you change the blur radius on the filter, the source seems to get larger, causing it to shift from wherever it was. Is there a way to fix that?

tjienta commented:
The source get larger because space needs to be made for the shadow which is rendered inside the source. Other filters that change the source size behave in the same way.

It would be nice if there was a way to specify and anchor point so that it doesn't feel as if the source is moving. But that would be a whole other can of worms that I don't want to open.

WizardCM commented:
The anchor point for a source by default is the Top Left. Right click -> Transform -> Edit Transform has a "Positional Alignment" dropdown. Change that to Center, and that'll stop the drop shadow shifting the source's position.

Xaymar commented:
@WizardCM I think the modification of the anchor point is not meant as the transform anchor point and much more as the actual relative render origin anchor point. Currently if you extend the source size in a filter it causes the source to "move" somewhere else. This is bad for drop shadow, glow, outline, etc, which all don't have to move the source.

WhiteExponent commented:
Any updates on this one? Seems to build and run fine on current master.

WizardCM commented:
Jim is in the process of going through PRs (77 merged since 24.0.3, with another 43 still open), and will get to this one when he can.

tjienta commented:
The drop shadow needs at maximum 7 passes. Jim thought at the time that it was too much GPU usage to include this in OBS.

It may be possible to change the algorithm to make a hybrid together with scaling down, then scaling up at the same time as sampling the image. This could reduce the amount of passes by at least a half. But it would be a bit of work to implement this and I do not currently have the time to pick up this task.

WhiteExponent commented:
If it helps, I am using this filter with 3 different layers in the same scene with a Radeon HD 5670 and it seems to output fine at 60FPS @720p (the scene has about 15 layers in total).

jp9000 commented:
I don't feel comfortable with this.  To be honest it's a lot of things.

I don't feel comfortable with the number of passes it uses.  Now while this is fine on most textures, the problem is that if people have really long text strings, it can allocate very large textures to accommodate that text because we don't currently have a way of cutting that off, and the number of passes will be significant affected by this.  Now that's mostly my fault.  Had I realized this sooner in program design I might have designed things a bit differently.

Secondly, one thing I'm starting to realize with the project is that there's a bit too much feature creep going on, and it's been a lot for me to handle.  I'd prefer to keep things focused and simple until I can get more maintenance help.

I'm not sure if there are ideal ways to compromise this for now, I think things would really need to be reworked to allow for this.  If anything, I'd say make this a third-party plugin or something for now rather than try to get it in the main project, at least for now.  My own personal burden is just a bit too much at the moment and it's stressing me out a lot.
